### PR TITLE
Update to std::future

### PR DESCRIPTION
**Pull request by seanmonstar**
This is a WIP branch of the update from `futures` v0.1 to `std::future::Future`. I'm posting it so the outside can track progress, and allow others to try to contribute. See #1805.

### Steps

- [x] Remove `futures` v0.1 from dependencies
- [x] Switch to `tokio` master that has `std::future` support
- [x] Get it compiling (at least the type signatures)
- [x] Fill in server implementation
  - [x] `examples/hello.rs` runs and sends responses
  - [x] Figure out "HTTP Upgrade" stuff
- [x] Fill in client implementation
  - [x] The `dispatch` channel
  - [x] The `Connect` trait
  - [x] Update the `Pool`
  - [x] `examples/client.rs` runs and prints the body
  - [x] Figure out "HTTP Upgrades"
- [x] `hyper::Body`
- [ ] Update `h2` support
- [ ] Audit new `unsafe` usage of `Pin` 😭

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -29,16 +33,18 @@ iovec = "0.1"
...
+tokio = { path = "../tokio/tokio", optional = true, default-features = false, features = ["rt-full"] }
 tokio-buf = "0.1"
-tokio-executor = { version = "0.1.0", optional = true }
-tokio-io = "0.1"
-tokio-reactor = { version = "0.1", optional = true }
-tokio-tcp = { version = "0.1", optional = true }
-tokio-threadpool = { version = "0.1.3", optional = true }
-tokio-timer = { version = "0.2", optional = true }
-want = "0.0.6"
+tokio-executor = { path = "../tokio/tokio-executor", optional = true }
```
Comment (LucioFranco) [inline Cargo.toml]: might wanna make these git deps so I can run this locally? 😄

Comment (seanmonstar) [reply Cargo.toml]: Oh right, haven't updated since it was merged to master XD

**Comment by Aaron1011** [comment]
@seanmonstar: I believe that working with `Pin` shouldn't actually require any unsafe code.

Looking at the unsafe code you've added, all of it seems to be to support pin projections for custom `Future` structs that you declare. However, none of these types actually rely on the immovability guarantees provided by `Pin` - they're just regular structs, which currently work without any of the `Pin` guarantees.

This means that you should be able to add an unconditional `Pin` impl for all of these types, and simply call `Pin::get_mut` on `self`. Since you don't have any `unsafe` code that relies on these structs being immovable (e.g. self-referential fields), this will be perfectly safe, and will greatly simply the implementation.

**Comment by seanmonstar** [comment]
Most things have been updated. Still lacking http2 (hitting those paths will find an `unimplemented!`). A bit of the unsafety has been removed, but there's still some in here that should get looked at some more.

Master is now tracking 0.13, so I'm going to disable tests and then merge this to master, and file issues to re-enable tests and examples. Hopefully those are smaller chunks that can be parallelized better.

**Comment by DoumanAsh** [comment]
@seanmonstar JFYI current master cannot be built without `runtime` feature

**Comment by seanmonstar** [comment]
@DoumanAsh woops, thanks for letting me know. Master should now be working (though the client doesn't quite have a way to set a custom executor...).

**Comment by danieleades** [comment]
looks like setting a custom executor requires passing in something that implements  tokio_executor::Executor. Should this be changed to futures::task::Spawn?

**Comment by seanmonstar** [comment]
I think not yet. In fact, both are kinda unstable. I might suggest we just take a closure instead of some 3rd party trait. We can discuss that in its own issue.

**Pull request merged**